### PR TITLE
feat: isUUID

### PR DESCRIPTION
**Pull request by Noswad123**

### CODE REVIEW

File: src/Checker.ts [hunk 1]
```diff
@@ -1,14 +1,30 @@
...
-	() => getRawGuard(checker),
-	() => getRawNegation(checker)
-);
+	) => NegatableGuard<FROM, TO, N> = (checker) => buildNegatable(
+		() => getRawGuard(checker),
+		() => getRawNegation(checker)
+	);
+
+const uuidChecker = (input: string) => {
+	if (!!input
```
Comment (mgabeler-lee-6rs) [inline src/Checker.ts]: You don't need the `!!input` bit -- an empty string will trivially fail the regex

File: src/Checker.ts [hunk 2]
```diff
@@ -1,14 +1,30 @@
...
+			&& /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/.test(input)
```
Comment (mgabeler-lee-6rs) [inline src/Checker.ts]: You should store this regex external to the checker so that it is cached, and isn't getting recompiled for each use.

File: src/Checker.ts [hunk 3]
```diff
@@ -1,14 +1,30 @@
...
+	) {
+		return 'Valid UUID';
```
Comment (mgabeler-lee-6rs) [inline src/Checker.ts]: There's a specific pattern used for the confirm/error strings to allow composition of them in arrays & property guards and the like, should stick to that.

File: src/Checker.ts [hunk 4]
```diff
@@ -1,14 +1,30 @@
...
-) => NegatableGuard<FROM, TO, N> = (checker) => buildNegatable(
-	() => getRawGuard(checker),
-	() => getRawNegation(checker)
-);
+	) => NegatableGuard<FROM, TO, N> = (checker) => buildNegatable(
+		() => getRawGuard(checker),
+		() => getRawNegation(checker)
+	);
+
+const uuidChecker = (input: string) => {
```
Comment (mgabeler-lee-6rs) [inline src/Checker.ts]: This belongs in `restrictingGuards`

That move will also fix your import dependency cycle

File: src/Utils.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const UUIDRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/;
```
Comment (mgabeler-lee-6rs) [inline src/Utils.ts]: This doesn't need to be exported, it just needed to be a const not re-instantiated for each guard invocation...

File: src/Checker.ts [hunk 6]
```diff
@@ -5,10 +5,10 @@ export type Checker<FROM> = (input: FROM) => string;
 
 export const checkerToGuard: <FROM, TO extends FROM, N extends FROM = FROM>(
 	checker: Checker<FROM>
-) => NegatableGuard<FROM, TO, N> = (checker) => buildNegatable(
-	() => getRawGuard(checker),
-	() => getRawNegation(checker)
-);
+	) => NegatableGuard<FROM, TO, N> = (checker) => buildNegatable(
```
Comment (mgabeler-lee-6rs) [inline src/Checker.ts]: revert the noise whitespace changes

File: src/restrictingGuards.ts [hunk 7]
```diff
@@ -2,16 +2,17 @@ import {checkerToGuard} from './Checker';
 import {andGuard, orGuard, thenGuard} from './Combinators';
 import {ReasonGuard} from './ReasonGuard';
 import {isNumber, isSymbol, isString} from './primitiveGuards';
-import {isDate} from './instanceGuards';
-import {isDateString} from './parseGuards';
+import {UUIDRegex} from './Utils';
 
-export const isDateOrDateString = orGuard(
```
Comment (mgabeler-lee-6rs) [inline src/restrictingGuards.ts]: pretty sure you didn't mean to delete this

File: src/restrictingGuards.ts [hunk 8]
```diff
@@ -13,6 +13,18 @@ export const isDateOrDateString = orGuard(
 	isDate
 );
 
+const UUIDRegex = /[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/;
```
Comment (mgabeler-lee-6rs) [inline src/restrictingGuards.ts]: You need `^...$` anchors on this, otherwise this will match any string that _contains_ a UUID

File: test/restrictingGuards.spec.ts [hunk 9]
```diff
@@ -16,6 +16,16 @@ describe('restricting guards', function() {
...
 		});
 	});
+	context('isUUID', function() {
+		it('should guard for a valid UUID', function() {
+			assertGuards(true)(restricting.isUUID, '0ca8f69c-1d07-4404-9b82-d1d0eb492313');
+		});
+		it('should guard against a invalid UUID', function() {
+			assertGuards(false)(restricting.isUUID, '');
+			assertGuards(false)(restricting.isUUID, '123');
+			assertGuards(false)(restricting.isUUID, 123);
```
Comment (mgabeler-lee-6rs) [inline test/restrictingGuards.spec.ts]: Add tests for UUIDs prefixed/suffixed/both by other characters, whitespace or otherwise

**Comment by mgabeler-lee-6rs** [comment]
Pedantic: this is `feat: ` not `fix: ` :wink:
